File: mobile/src/main/java/mycroft/ai/MycroftApplication.java
package mycroft.ai;

import android.app.Application;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.TaskStackBuilder;
import android.content.Context;
import android.content.Intent;
import android.support.v4.app.NotificationCompat;
import android.util.Log;

import org.altbeacon.beacon.BeaconManager;
import org.altbeacon.beacon.BeaconParser;
import org.altbeacon.beacon.Region;
import org.altbeacon.beacon.powersave.BackgroundPowerSaver;
import org.altbeacon.beacon.startup.BootstrapNotifier;
import org.altbeacon.beacon.startup.RegionBootstrap;

/**
 * Created by sarahkraynick on 2017-07-29.
 */
public class MycroftApplication extends Application implements BootstrapNotifier {

    private static final String TAG = "MycroftApplication";
    private static Context context;
    private BeaconActivity beaconActivity;
    private RegionBootstrap regionBootstrap;
    private BackgroundPowerSaver backgroundPowerSaver;
    private boolean haveDetectedBeaconsSinceBoot = false;
    private BeaconManager beaconManager;

    public void onCreate() {
        super.onCreate();
        MycroftApplication.context = getApplicationContext();
        beaconManager = org.altbeacon.beacon.BeaconManager.getInstanceForApplication(this);

        Log.d(TAG, "setting up background monitoring for beacons and power saving");
        // wake up the app when a beacon is seen
        Region region = new Region("backgroundRegion",
                null, null, null);
        regionBootstrap = new RegionBootstrap(this, region);

        backgroundPowerSaver = new BackgroundPowerSaver(this);
        setBeaconScanSettings(getString(R.string.beacon_layout));
    }

    public static Context getAppContext() {
        return MycroftApplication.context;
    }

    /**
     * Set the beacon settings for the beacon layout. The are mapped in the string array.
     */
    public void setBeaconScanSettings(String beaconLayout) {
        beaconManager.getBeaconParsers().add(new BeaconParser().
                setBeaconLayout(beaconLayout));
    }

    public void setMonitoringActivity(BeaconActivity beaconActivity) {
        this.beaconActivity = beaconActivity;
    }

    @Override
    public void didEnterRegion(Region arg0) {
        Log.d(TAG, "did enter region.");
        if (!haveDetectedBeaconsSinceBoot) {
            Log.d(TAG, "auto launching MainActivity");

            Intent intent = new Intent(this, BeaconActivity.class);
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

            this.startActivity(intent);
            haveDetectedBeaconsSinceBoot = true;
        } else {
            if (beaconActivity != null) {
                // If the Monitoring Activity is visible, we log info about the beacons we have
                // seen on its display
                beaconActivity.logToDisplay("I see a beacon again" );
            } else {
                Log.d(TAG, "Sending notification.");
                sendNotification();
            }
        }
    }

    @Override
    public void didExitRegion(Region region) {
        if (beaconActivity != null) {
            beaconActivity.logToDisplay("I no longer see a beacon.");
        }
    }

    @Override
    public void didDetermineStateForRegion(int state, Region region) {
        if (beaconActivity != null) {
            beaconActivity.logToDisplay("I have just switched from seeing/not seeing beacons: " + state);
        }
    }

    private void sendNotification() {
        NotificationCompat.Builder builder =
                new NotificationCompat.Builder(this)
                        .setContentTitle("Mycroft Beacon")
                        .setContentText("An beacon is nearby.")
                        .setSmallIcon(R.drawable.common_plus_signin_btn_icon_dark);

        TaskStackBuilder stackBuilder = TaskStackBuilder.create(this);
        stackBuilder.addNextIntent(new Intent(this, BeaconActivity.class));
        PendingIntent resultPendingIntent =
                stackBuilder.getPendingIntent(
                        0,
                        PendingIntent.FLAG_UPDATE_CURRENT
                );
        builder.setContentIntent(resultPendingIntent);
        NotificationManager notificationManager =
                (NotificationManager) this.getSystemService(Context.NOTIFICATION_SERVICE);
        notificationManager.notify(1, builder.build());
    }
}

File: mobile/src/main/java/mycroft/ai/MycroftApplication.kt
package mycroft.ai

import android.app.Application
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.TaskStackBuilder
import android.content.Context
import android.content.Intent
import android.support.v4.app.NotificationCompat
import android.util.Log

import org.altbeacon.beacon.BeaconManager
import org.altbeacon.beacon.BeaconParser
import org.altbeacon.beacon.Region
import org.altbeacon.beacon.powersave.BackgroundPowerSaver
import org.altbeacon.beacon.startup.BootstrapNotifier
import org.altbeacon.beacon.startup.RegionBootstrap

/**
 * Created by sarahkraynick on 2017-07-29.
 */
class MycroftApplication : Application(), BootstrapNotifier {
    private var beaconActivity: BeaconActivity? = null
    private var regionBootstrap: RegionBootstrap? = null
    private var backgroundPowerSaver: BackgroundPowerSaver? = null
    private var haveDetectedBeaconsSinceBoot = false
    private var beaconManager: BeaconManager? = null

    override fun onCreate() {
        super.onCreate()
        MycroftApplication.appContext = applicationContext
        beaconManager = org.altbeacon.beacon.BeaconManager.getInstanceForApplication(this)

        Log.d(TAG, "setting up background monitoring for beacons and power saving")
        // wake up the app when a beacon is seen
        val region = Region("backgroundRegion", null, null, null)
        regionBootstrap = RegionBootstrap(this, region)

        backgroundPowerSaver = BackgroundPowerSaver(this)
        setBeaconScanSettings(getString(R.string.beacon_layout))
    }

    /**
     * Set the beacon settings for the beacon layout. The are mapped in the string array.
     */
    fun setBeaconScanSettings(beaconLayout: String) {
        beaconManager!!.beaconParsers.add(BeaconParser().setBeaconLayout(beaconLayout))
    }

    fun setBeaconActivity(beaconActivity: BeaconActivity) {
        this.beaconActivity = beaconActivity
    }

    fun unsetBeaconActivity() {
        this.beaconActivity = null
    }

    override fun didEnterRegion(arg0: Region) {
        Log.d(TAG, "did enter region.")
        if (!haveDetectedBeaconsSinceBoot) {
            Log.d(TAG, "auto launching MainActivity")

            val intent = Intent(this, BeaconActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK

            this.startActivity(intent)
            haveDetectedBeaconsSinceBoot = true
        } else {
            if (beaconActivity != null) {
                // If the Monitoring Activity is visible, we log info about the beacons we have
                // seen on its display
                beaconActivity!!.logToDisplay("I see a beacon again")
            } else {
                Log.d(TAG, "Sending notification.")
                sendNotification()
            }
        }
    }

    override fun didExitRegion(region: Region) {
        if (beaconActivity != null) {
            beaconActivity!!.logToDisplay("I no longer see a beacon.")
        }
    }

    override fun didDetermineStateForRegion(state: Int, region: Region) {
        if (beaconActivity != null) {
            beaconActivity!!.logToDisplay("I have just switched from seeing/not seeing beacons: $state")
        }
    }

    private fun sendNotification() {
        val builder = NotificationCompat.Builder(this)
                .setContentTitle(getString(R.string.beacon_notification_title))
                .setContentText(getString(R.string.beacon_notification_content))
                .setSmallIcon(R.drawable.ic_mycroft)

        val stackBuilder = TaskStackBuilder.create(this)
        stackBuilder.addNextIntent(Intent(this, BeaconActivity::class.java))
        val resultPendingIntent = stackBuilder.getPendingIntent(
                0,
                PendingIntent.FLAG_UPDATE_CURRENT
        )
        builder.setContentIntent(resultPendingIntent)
        val notificationManager = this.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        notificationManager.notify(1, builder.build())
    }

    //TODO convert to public member. This will cause problems later.
    companion object {

        private val TAG = "MycroftApplication"
        var appContext: Context? = null
            private set
    }
}
